src/parser/jobs/smn/modules/Ruin4.tsx

```diff
@@ -0,0 +1,137 @@
+import {t} from '@lingui/macro'
+import {Plural, Trans} from '@lingui/react'
+import {ActionLink} from 'components/ui/DbLink'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {CastEvent} from 'fflogs'
+import Module, {dependency} from 'parser/core/Module'
+import Invulnerability from 'parser/core/modules/Invulnerability'
+import Suggestions, {SEVERITY, Suggestion, TieredSuggestion} from 'parser/core/modules/Suggestions'
+import React from 'react'
+
+const SEVERITY_STACK_COUNT = {
+	1: SEVERITY.MEDIUM,
+	5: SEVERITY.MAJOR,
+}
+
+const FURTHER_RUIN_PLAYER_ACTIONS = [
+	ACTIONS.ASSAULT_I_AERIAL_SLASH.id,
+	ACTIONS.ASSAULT_II_SLIIPSTREAM.id,
+	ACTIONS.ASSAULT_I_CRIMSON_CYCLONE.id,
+	ACTIONS.ASSAULT_II_FLAMING_CRUSH.id,
+	// NOTE: Titan Egi Assualt I Earthen Armor does not generate a Further Ruin stack
+	ACTIONS.ASSAULT_II_MOUNTAIN_BUSTER.id,
+]
+
+const FURTHER_RUIN_PET_ACTIONS = [
+	ACTIONS.AERIAL_SLASH.id,
+	ACTIONS.SLIPSTREAM.id,
+	ACTIONS.CRIMSON_CYCLONE.id,
+	ACTIONS.FLAMING_CRUSH.id,
+	// NOTE: Titan Egi Assualt I does not generate a Further Ruin stack
+	ACTIONS.MOUNTAIN_BUSTER.id,
+]
+
+const MAX_FURTHER_RUIN_COUNT = 4
+const END_OF_FIGHT_LEEWAY = 5000
+
+export default class Ruin4 extends Module {
+	static handle = 'ruin4'
+	static title = t('smn.ruin-iv.title')`Ruin IV`
+
+	@dependency private suggestions!: Suggestions
+	@dependency private invuln!: Invulnerability
+
+	private currentStackCount = 0
+	private earthenArmorCount = 0
+	private overage = 0
+	private playerSkillCount = 0
+	private petSkillCount = 0
+
+	protected init() {
+		this.addHook('cast', {by: 'player'}, this.onPlayerCast)
+		this.addHook('cast', {by: 'pet'}, this.onPetCast)
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onPlayerCast(event: CastEvent) {
+		if (event.ability.guid === ACTIONS.RUIN_IV.id) {
```

Comment:
not blocking; but each of these conditional blocks could be implemented as hook filters